modules/substrate/src/lib.rs

```diff
@@ -31,43 +31,95 @@
 // Runtime-generated enums
 #![allow(clippy::large_enum_variant)]
 
-use bp_substrate::{AuthoritySet, ImportedHeader, ScheduledChange};
+use crate::storage::{AuthoritySet, ImportedHeader, ScheduledChange};
+use codec::{Codec, EncodeLike};
 use frame_support::{decl_error, decl_module, decl_storage, dispatch};
 use frame_system::ensure_signed;
-use sp_runtime::traits::Header as HeaderT;
-use sp_std::{marker::PhantomData, prelude::*};
+use num_traits::AsPrimitive;
+use sp_runtime::traits::{
+	AtLeast32BitUnsigned, Hash as HashT, Header as HeaderT, MaybeDisplay, MaybeMallocSizeOf, MaybeSerializeDeserialize,
+	Member, SimpleBitOps,
+};
+use sp_std::{fmt::Debug, marker::PhantomData, prelude::*, str::FromStr};
 
 mod justification;
+mod storage;
 mod storage_proof;
 mod verifier;
 
 #[cfg(test)]
 mod mock;
 
-type Hash<T> = <T as HeaderT>::Hash;
-type Number<T> = <T as HeaderT>::Number;
-
-pub trait Trait: frame_system::Trait {}
+pub trait Trait: frame_system::Trait {
+	/// A type that fulfulls the abstract idea of what a Substrate header is.
+	// See here for more info:
+	// https://crates.parity.io/sp_runtime/traits/trait.Header.html
+	type BridgedHeader: HeaderT<Number = Self::BridgedBlockNumber, Hash = Self::BridgedBlockHash>;
+
+	/// A type that fulfills the abstract idea of what a Substrate block number is.
+	// Constraits come from the associated Number type of `sp_runtime::traits::Header`
+	// See here for more info:
+	// https://crates.parity.io/sp_runtime/traits/trait.Header.html#associatedtype.Number
+	//
+	// Note that the `AsPrimitive<usize>` trait is required by the Grandpa justification
+	// verifier, and is not usually part of a Substrate Header's Number type.
+	type BridgedBlockNumber: Member
+		+ MaybeSerializeDeserialize
+		+ Debug
+		+ sp_std::hash::Hash
+		+ Copy
+		+ MaybeDisplay
+		+ AtLeast32BitUnsigned
+		+ Codec
+		+ FromStr
+		+ MaybeMallocSizeOf
+		+ AsPrimitive<usize>;
+
+	/// A type that fulfills the abstract idea of what a Substrate hash is.
+	// Constraits come from the associated Hash type of `sp_runtime::traits::Header`
+	// See here for more info:
+	// https://crates.parity.io/sp_runtime/traits/trait.Header.html#associatedtype.Hash
+	type BridgedBlockHash: Member
+		+ MaybeSerializeDeserialize
+		+ Debug
+		+ sp_std::hash::Hash
+		+ Ord
+		+ Copy
+		+ MaybeDisplay
+		+ Default
+		+ SimpleBitOps
+		+ Codec
+		+ AsRef<[u8]>
+		+ AsMut<[u8]>
+		+ MaybeMallocSizeOf
```

Comment:
A bunch of these traits is already part of `Member`. There is also `Parameter` trait in the substrate repo, do we really need to list all of them here?

Reply:
I wasn't sure if there was already a trait in Substrate that captured this. I think Slava did a similar thing in a recent relay refactor PR. I can take a look at `Member` and `Parameter` though

Reply:
While we can clean up _some_ trait bounds which are duplicated in both `Member` and `Parameter` I don't think it's worth it. We explicitly point the reader to the trait bounds in `sp_runtime::traits::Header` since that's where they were taken from and we should be consistent with that.

In the future though we definitely should move all the traits bounds to a more central location where we can hide it behind a single trait.

Reply:
`sp_runtime::traits::Header` just has excessive bounds, probably no one really took a look at cleaning them up at some point, when some of the traits got merged into `Member`.

IMHO repeating the same bounds multiple time is just super dumb (basically like writing `Clone + Clone`), and the fact that they are repeated in substrate codebase is not a compelling argument for that either. That said I'd rather get the PR merged than go back and forth about it:)